Add new registration email for new users.

LinusFresz opened the pull request:
As a follow-up to https://github.com/thewca/worldcubeassociation.org/issues/1814 I've looked into creating a new confirmation mail for newly created user accounts.

However, we can only change the reconfirmation instructions mail (which was originally requested in https://github.com/thewca/worldcubeassociation.org/issues/1814 if we overwrite `send_reconfirmation_instructions` in our User model as well. To be honest, I am not entirely sure if it's worth it to only change a few words.

I've already added tests (which are going to fail for the foreign language mails) to test the new registration mail. I used https://github.com/thewca/worldcubeassociation.org/blob/f66b258c7d8f54defc0544d912c6d521f99fc316/WcaOnRails/spec/mailers/registrations_mailer_spec.rb Lines 26 - 46 as comparison to create a test for foreign languages as well, but the test for competition registration mails obviously assumes that we have translations for the confirmation mails. How can I avoid to run into test errors for these language tests while we only have a user registration confirmation mail in english?

I am definitely open for discussions about the mail text! :)

Screenshot:
![new-registration-mail](https://user-images.githubusercontent.com/18479675/41935411-e5af5122-7989-11e8-9284-6c5dcd215cb6.png)

jfly commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -602,6 +602,15 @@ en:
+        confirm: "Your account is created and must be activated before you can use it. Please confirm your account through the link below: "
+        activate: "After activation you may login to the website of the World Cube Association: "
+        addendum: "This email is generated automatically. If you didn't register an account on the website of the World Cube Association and received this message by mistake, please ignore and delete it."
+        questions: "Do you have more questions? We will be happy to answer them! Feel free to send a mail to "
```
Translations like this are always concerning to me, because it assumes that the language is going to want this link or whatever to come at the end of the sentence. We prefer to use string interpolation whenever possible, so this would become something like:

```
questions: "Do you have more questions? We will be happy to answer them! Feel free to send a mail to %{contact_link}."
```

After this change, you would also have to change view code a bit.

I think you could make this same sort of change for the `activate` rule above.

jfly commented on WcaOnRails/spec/mailers/new_registration_mailer_spec.rb:
```diff
@@ -0,0 +1,47 @@
+    end
+  end
+
+  describe "send registration mail to new users in their language" do
```
I don't think it's terribly important to test this. As you already mentioned, these tests won't actually pass until the French translations are done anyways.

This sort of test is from the early days of internationalization on the WCA website when we were still trying to figure out how to do it right (that `localized_mail` method you're using is what we ended up building to make this easier on us).

LinusFresz replied:
Alright, thanks for letting me know! I just hoped there is some way to check if either a translation is available or use the English text instead. But as it's not, I'll remove the translation test.

jfly commented on WcaOnRails/app/models/user.rb:
```diff
@@ -331,6 +331,10 @@ def avatar_requires_wca_id
     end
   end
 
+  def send_on_create_confirmation_instructions
```
I think it would be good to have a test that confirms that this method actually gets called when a user registers. I think that would look like a test in `user_spec` that mocks out `NewRegistrationMailer.send_registration_mail` and asserts that it is called with this new user.

LinusFresz replied:
Fixed tests and mail text!

I also created all necessary files for the new email address change confirmation mail as requested in https://github.com/thewca/worldcubeassociation.org/issues/1814. 

Screenshot:
![new-mail-confirmation](https://user-images.githubusercontent.com/18479675/42052431-1a64eb4e-7b0f-11e8-9cb8-34f2b19add2b.png)

LinusFresz commented:
sorry for the still failing tests, I am going to fix them tomorrow.

LinusFresz commented:
Ok, so tests are fixed now and travis build passed! However it says "Coverage decreased", what exactly does it mean? Do I have to take any action?

jfly commented on WcaOnRails/spec/models/user_spec.rb:
```diff
@@ -353,6 +353,27 @@
+  describe "check if registration form sends mail to newly registered user" do
+    it "send mail" do
+      new_user = FactoryBot.create :user, :unconfirmed
+      expect(ActionMailer::Base.deliveries.last.to).to eq [new_user.email]
```
This is an interesting way of checking if the email was sent! However, we haven't been checking for emails this way, as it's a little brittle. You already tested that the delivery address and subject are correct in your mailer spec. Instead, we've been using mocks. If you look in this file, you'll see https://github.com/jfly/worldcubeassociation.org/blob/ebb8e956eef9ea57b112333f8ba5835ba0193731/WcaOnRails/spec/models/user_spec.rb#L498, where we check that the right mailer method was called with the correct arguments.

jfly commented on WcaOnRails/spec/models/user_spec.rb:
```diff
@@ -353,6 +353,27 @@
+      user.send_reconfirmation_instructions
+
+      expect { user.send_reconfirmation_instructions } .to change { ActionMailer::Base.deliveries.count }.by(1)
+      expect(ActionMailer::Base.deliveries.last.to).to eq [user.unconfirmed_email]
```
As I commented above, I think it would be better to check a mock here, rather than looking at the actual email that was sent.

jfly commented on WcaOnRails/spec/models/user_spec.rb:
```diff
@@ -353,6 +353,27 @@
+      user = FactoryBot.create :user
+      user.unconfirmed_email = "new@example.com"
+      expect(user.pending_reconfirmation?).to eq true
+      user.send_reconfirmation_instructions
```
Is this really supposed to be here? I'm surprised because you call the same method on the next line inside of the `expect` block.

LinusFresz replied:
Woops, missed this one from my previous work, will simply remove this line.

jfly commented on WcaOnRails/spec/models/user_spec.rb:
```diff
@@ -353,6 +353,27 @@
+      expect(user.pending_reconfirmation?).to eq true
+      user.send_reconfirmation_instructions
+
+      expect { user.send_reconfirmation_instructions } .to change { ActionMailer::Base.deliveries.count }.by(1)
```
Instead of explicitly calling `user.send_reconfirmation_instructions`, can you instead do a `user.save!` here? The problem with the test as currently written is that it checks that the `send_reconfirmation_instructions` method does the right thing, but it doesn't check that devise actually calles the `send_reconfirmation_instructions` method.

jfly commented on WcaOnRails/app/views/new_registration_mailer/send_registration_mail.html.erb:
```diff
@@ -0,0 +1,25 @@
+<p>
+  <%= t 'users.mailer.create_new_account.activate' %>
+
+  <%= link_to('login form', 'https://www.worldcubeassociation.org/users/sign_in') %>
```
nit: This won't work for environments with a different url (such as staging or local development). It would be better to use `new_user_registration_url` here instead.

jfly commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -602,6 +602,23 @@ en:
+        welcome: "Welcome %{user_email}!"
+        confirm: "Your account is created and must be activated before you can use it. Please confirm your account through the link below: "
+        confirmation_link: "Confirm my account."
+        activate: "After activation you may login to the website of the World Cube Association: "
```
nit: login -> log in. "login" is a noun, "log in" is the verb phrase.

jfly commented:
> However it says "Coverage decreased", what exactly does it mean? Do I have to take any action?

We use a tool to keep track of how much of our real code is actually executed when we run our tests. The idea is that if we write a whole bunch of code, and none of our tests actually execute that code, that's a sign that new code is probably untested and could easily break as future people work on it.

I think your test coverage here is pretty good. I'm not sure why exactly the coverage when down by 0.01%, but I think it's ok to ignore here.

jfly commented on WcaOnRails/app/models/user.rb:
```diff
@@ -331,6 +331,18 @@ def avatar_requires_wca_id
+    NewRegistrationMailer.send_registration_mail(self).deliver_now
+  end
+
+  def send_reconfirmation_instructions
```
Can you add a comment here about how this was copied from https://github.com/plataformatec/devise/blob/master/lib/devise/models/confirmable.rb#L124?

LinusFresz replied:
Sure, will do!

jfly commented on WcaOnRails/app/models/user.rb:
```diff
@@ -331,6 +331,18 @@ def avatar_requires_wca_id
+    @reconfirmation_required = false
+
+    unless @skip_confirmation_notification
+      ReconfirmationMailer.send_reconfirmation_mail(self).deliver_now
```
Actually, I'm concerned that this is going to break reconfirmation. This line of code looks a lot different than the `def send_confirmation_instructions` https://github.com/plataformatec/devise/blob/b8f644198d34e96039b84b69842ecb4edffa4664/lib/devise/models/confirmable.rb#L115-L122 that is currently happening.

Can we simply not override this method, and instead edit WcaOnRails/app/views/devise/mailer/confirmation_instructions.html.erb?

LinusFresz replied:
the lines of code you are talking about are actually `send_*re*confirmation_instructions`, not `send_confirmation_instructions`. And besides the changed method call, the methods are identical! I did not change anything about `send_confirmation_instructions`, so we're still using Devise's method for this, but `send_reconfirmation_instructions` does not call `send_confirmation_instructions` anymore but the newly created mailer instead.

jfly replied:
Sorry, my comment wasn't very clear. Let me try again:

The method you've written here looks identical to devise's, **except** for the fact that devises's calls `send_confirmation_instructions`, and yours calls `ReconfirmationMailer.send_reconfirmation_mail(self).deliver_now`. I'm concerned that this is not a safe change, because `send_confirmation_instructions` looks like it does a number of things (such as sometimes generating a confirmation token?), and the new code you've written doesn't do that.

It doesn't feel like this `send_reconfirmation_instructions` method is supposed to be overridden.

LinusFresz commented:
I've added some fixes! Most importantly I removed `send_reconfirmation_instructions` and used the 'confirmation mail' template from Devise for the email reconfirmation.

Note, that I've changed the `:unconfirmed` trait into a transient `confirmed` in users.rb to avoid automatic confirmation if needed (as for my case). Only one other test seemed to used this trait and should be fixed as well.

LinusFresz commented on WcaOnRails/app/models/user.rb:
```diff
@@ -331,6 +331,11 @@ def avatar_requires_wca_id
     end
   end
 
+  # The methods send_on_create_confirmation_instructions and send_reconfirmation_instructions were copied from https://github.com/plataformatec/devise/blob/master/lib/devise/models/confirmable.rb
```
well, technically `send_on_create_confirmation instructions` is still overridden from Devise, so I am going to edit it.

jfly commented:
Awesome work here, @LinusFresz!
